### PR TITLE
Finish bulk memory support

### PR DESCRIPTION
**Pull request by tlively**
Implement interpretation of remaining bulk memory ops, add bulk memory
spec tests with light modifications, fix bugs preventing the fuzzer
from running correctly with bulk memory, and fix bugs found by the
fuzzer.

### CODE REVIEW

File: src/wasm-interpreter.h [hunk 1]
```diff
@@ -1253,12 +1256,60 @@ class ModuleInstanceBase {
...
+      Flow size = this->visit(curr->size);
+      if (size.breaking()) return size;
+      NOTE_EVAL1(dest);
+      NOTE_EVAL1(source);
+      NOTE_EVAL1(size);
+      Address destVal(uint32_t(dest.value.geti32()));
+      Address sourceVal(uint32_t(source.value.geti32()));
+      Address sizeVal(uint32_t(size.value.geti32()));
+
+      instance.checkLoadAddress(destVal, 0);
```
Comment (binji) [inline src/wasm-interpreter.h]: This doesn't seem right to me -- the bulk memory proposal requires that a copy that would trap is partially written, up to the point that the trap occurs.

Comment (tlively) [reply src/wasm-interpreter.h]: Yes, but this line is checking that the destination address is in bounds at all (or one past the end, as we discussed earlier). The bounds checking for each individual byte is done by `getFinalAddress` in the loop below.

Comment (binji) [reply src/wasm-interpreter.h]: Oh, does that throw an exception? If so that works too.

File: src/passes/RemoveUnusedModuleElements.cpp [hunk 2]
```diff
@@ -119,6 +119,18 @@ struct ReachabilityAnalyzer : public PostWalker<ReachabilityAnalyzer> {
...
+  void visitDataDrop(DataDrop* curr) {
+    usesMemory = true;
+  }
+  void visitMemoryCopy(MemoryCopy* curr) {
+    usesMemory = true;
+  }
+  void visitMemoryFill(MemoryFill* curr) {
+    usesMemory = true;
+  }
   void visitHost(Host* curr) {
```
Comment (kripken) [inline src/passes/RemoveUnusedModuleElements.cpp]: we should probably refactor this pass to just use Effects to see if memory is used. Please add a TODO.

File: src/tools/wasm-opt.cpp [hunk 3]
```diff
@@ -240,6 +240,7 @@ int main(int argc, const char* argv[]) {
     auto input = buffer.getAsChars();
     WasmBinaryBuilder parser(other, input, false);
     parser.read();
+    options.applyFeatures(other);
```
Comment (kripken) [inline src/tools/wasm-opt.cpp]: why is this needed / how did things work without it before? :)

File: src/wasm/wasm-binary.cpp [hunk 4]
```diff
@@ -1540,7 +1540,10 @@ void WasmBinaryBuilder::readDataSegments() {
     }
     curr.isPassive = flags & BinaryConsts::IsPassive;
     if (flags & BinaryConsts::HasMemIndex) {
-      curr.index = getU32LEB();
+      uint32_t memIndex = getU32LEB();
```
Comment (kripken) [inline src/wasm/wasm-binary.cpp]: `auto`

File: src/tools/wasm-opt.cpp (hunk 3 above)
Comment (tlively) [reply src/tools/wasm-opt.cpp]: This is in the fuzzing code, so I guess we hadn't tried fuzzing with features recently.
